File: java/main/src/pl/vot/tomekby/World/World.java
package pl.vot.tomekby.World;
import javafx.util.Pair;
import pl.vot.tomekby.World.Misc.Controller;
import pl.vot.tomekby.World.Misc.NoPlaceToSpawn;

import java.util.Comparator;
import java.util.Random;
import java.util.UUID;
import java.util.Vector;

/**
 * Created by Tomasz on 2015-05-02.
 */
final public class World {
    public World() { setRand_class(new Random()); }
    public void nextRound() {
        // Wykonywanie akcji organizmów
        for(Organism o : getOrganisms()) {
            // Jeśli nie oznaczony do usunięcia
            if(to_remove.indexOf(o.get_uid()) == -1) o.action();
        }
        for (int i = 0; i < getOrganisms().size(); ++i) {
            for (int j = i + 1; j < getOrganisms().size(); ++j) {
                // Te same koordynaty
                try {
                    if (getOrganisms().get(j).pos().equals(getOrganisms().get(i).pos()) && !getOrganisms().get(i).get_uid().equals(getOrganisms().get(j).get_uid())) {
                        getOrganisms().get(i).collision_default(getOrganisms().get(j));
                    }
                } catch(NoPlaceToSpawn e) {
                    System.err.println(e.getMessage());
                }
            }
        }
        // Kasowanie niepotrzebnych organizmów
        remove_pending();
        // Rysowanie świata
        draw_world();
    }
    public void add_organism(Organism new_org) {
        if (is_free((Integer) new_org.pos().getKey(), (Integer) new_org.pos().getValue())) {
            new_org.set_world(this);
            to_add.add(new_org);
        }
    }
    public void add_pending() {
        // Dopisanie zespawnowanych
        for(Organism o : to_add) {
            boolean ommit = false;
            for(Organism o2 : getOrganisms()) if(o.pos().equals(o2.pos())) ommit = true;
            // Jeśli pole jest wolne, dodanie zwierzęcia
            if(!ommit) getOrganisms().add(o);
        }
        to_add.clear();

        // Sortowanie w odpowiedniej kolejności
        getOrganisms().sort(new OrgComp());
    }
    public boolean is_free(int x, int y) {
        if (x < 0 || x >= N || y < 0 || y >= N) return false;

        final Pair pos = new Pair(x, y);
        for(Organism o : getOrganisms()) if(o.pos().equals(pos)) return false;
        for(Organism o : to_add) if(o.pos().equals(pos)) return false;
        return true;
    }
    public int get_N() { return N; }
    public void set_N(int N) { this.N = N; }
    public void add_event(String msg) {
        String entry = (log.size() + 1)+". "+msg;
        log.add(entry);
    }

    public Vector<Organism> getOrganisms() {
        return organisms;
    }

    public Vector<UUID> getTo_remove() {
        return to_remove;
    }

    public Vector<Organism> getTo_add() {
        return to_add;
    }

    // Porównywanie organizmów
    static class OrgComp implements Comparator<Organism> {
        public int compare(Organism o1, Organism o2) {
            if(o1.initiative != o2.initiative) return o2.initiative - o1.initiative;
            else if(o1.creation == o2.creation) return 0;
            return (int)(o1.creation - o2.creation);
        }
    }
    public void draw_world() {
        // Porządki
        controller.clear_info();
        // Czyszczenie mapki
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < N; ++j) controller.print(i, j, "EMPTY");
        }

        // Rysowanie organizmów
        draw_organisms();
        // Wypisywanie instrukcji
        print_instruction();
        // Wypisywanie dziennika zdarzeń
        controller.print_info("Dziennik zdarzeń:");
        for(String msg : log) controller.print_info(msg);
        log.clear();

        // Dopisanie zespawnowanych
        add_pending();
    }

    public void kill(Organism winner, Organism looser) {
        add_event(winner.get_name() + " zabija " + looser.get_name());
        to_remove.add(looser.get_uid());
    }
    public void eat(Organism a, Organism p) {
        add_event(a.get_name() + " zjada " + p.get_name());
        to_remove.add(p.get_uid());
    }
    private void remove_pending() {
        for(UUID id : to_remove) {
            for(Organism o : getOrganisms()) {
                if(o.get_uid().equals(id)) {
                    getOrganisms().remove(o);
                    break;
                }
            }
        }
        // Wyczyszczenie listy do usunięcia
        to_remove.clear();
    }
    private void draw_organisms() {
        for(Organism o : getOrganisms()) o.draw();
    }
    private void print_instruction() {
        String[] instruction = {"Strzałki - poruszanie człowiekiem", "Spacja - umiejętność specjalna człowieka", "Plik z zapisem stanu gry: world.xml"};
        for(String s : instruction) {
            controller.print_info(s);
        }
        controller.print_info("\n");
    }

    private Vector<Organism> organisms = new Vector<Organism>();
    private Vector<UUID> to_remove = new Vector<UUID>(); // UID organizmów do skasowania
    private Vector<Organism> to_add = new Vector<Organism>(); // Lista organizmów do zespawnowania
    private Vector<String> log = new Vector<String>(); // Lista zdarzeń
    private int N = 20;

    private Random rand_class;
    private Controller controller;

    public Random getRand_class() {
        return rand_class;
    }

    public void setRand_class(Random rand_class) {
        this.rand_class = rand_class;
    }

    public Controller getController() {
        return controller;
    }

    public void setController(Controller controller) {
        this.controller = controller;
    }
}

File: py/World/World.py
from Misc import Controller, CantSpawnException


class World(object):
    def __init__(self, rand_class=None, controller_class=None):
        """
        :type rand_class: random.randint
        :type controller_class: Controller
        :return:
        """
        self._rand = rand_class
        self._controller = controller_class

    # Wykonywanie następnej rundy
    def nextRound(self):
        for o in self._organisms:
            if not o.get_uid() in self._to_remove:
                o.action()
        # Sprawdzanie kolizji
        for i, org in enumerate(self._organisms):
            for j, other in enumerate(self._organisms[i + 1:]):
                if self._organisms[j].pos() == self._organisms[i].pos() and self._organisms[j] != self._organisms[i]:
                    try:
                        self._organisms[i].collision_default(self._organisms[j])
                    except CantSpawnException:
                        print("Nie można utworzyć organizmu")
        # Kasowanie niepotrzebnych + rysowanie świata
        self._remove_pending()
        self.draw_world()

    # Dodawanie nowego organizmu do świata
    def add_organism(self, new_org):
        if self.is_free(new_org.pos()):
            new_org.set_world(self)
            self._to_add.append(new_org)

    # Dodawanie oczekujących zwierząt
    def add_pending(self):
        for o in self._to_add:
            ommit = False
            for o2 in self._organisms:
                if o.pos() == o2.pos():
                    ommit = True
            if not ommit:
                self._organisms.append(o)
        self._to_add.clear()
        # Sortowanie aby zachować odpowiednią kolejność
        self._organisms.sort()

    # Czy dane pole jest puste?
    def is_free(self, pos: tuple):
        x, y = pos
        if x < 0 or x >= self._N or y < 0 or y >= self._N:
            return False

        for o in self._organisms:
            if o.pos() == pos:
                return False
        for o in self._to_add:
            if o.pos() == pos:
                return False
        return True

    # Dodawanie wpisu do dziennika
    def add_event(self, msg):
        self._log.append("%d. %s" % (len(self._log) + 1, msg))

    # Rysowanie świata
    def draw_world(self):
        # Porządki
        self._controller.clear_info()
        self._controller.reset_map()

        self._draw_organisms()
        self._print_instruction()
        self._controller.print_info("Dziennik zdarzeń:")
        for msg in self._log:
            self._controller.print_info(msg)
        self._log.clear()

        self.add_pending()

    # Zabijanie organizmu
    def kill(self, winner, looser):
        self.add_event(winner.get_name() + " zabija " + looser.get_name())
        self._to_remove.append(looser.get_uid())

    # Zjadanie organizmu
    def eat(self, animal, plant):
        self.add_event(animal.get_name() + " zjada " + plant.get_name())
        self._to_remove.append(plant.get_uid())

    # Usuwanie oczekujących organizmów
    def _remove_pending(self):
        for id in self._to_remove:
            for o in self._organisms:
                if o.get_uid() == id:
                    self._organisms.remove(o)
                    break
        self._to_remove.clear()

    # Rysowanie organizmów
    def _draw_organisms(self):
        for o in self._organisms:
            o.draw()

    # Wypisanie instrukcji do gry
    def _print_instruction(self):
        instruction = ["Strzałki - poruszanie człowiekiem", "Spacja - umiejętność specjalna człowieka"]
        for s in instruction:
            self._controller.print_info(s)
        self._controller.print_info("\n")

    # Pobieranie klasy kontrolera
    def get_controller(self) -> Controller:
        return self._controller

    # Ustawianie klasy kontrolera
    def set_controller(self, controller: Controller):
        self._controller = controller

    def getN(self):
        return self._N

    def setN(self, n):
        self._N = n

    # Zmienne
    _organisms = []
    _to_remove = []
    _to_add = []
    _log = []
    _N = 20
    # Klasy używane w dependency injection
    _rand = None
    _controller = None
